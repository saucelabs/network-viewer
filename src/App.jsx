import React from 'react';
import { NetworkProvider } from './state/network/provider';
import FilterContainer from './Containers/FilterContainer';
import Notification from './Components/Notification';
import NetworkTableContainer from './Containers/NetworkTableContainer';

function App() {
  return (
    <NetworkProvider>
      <Notification />
      <FilterContainer />
      <NetworkTableContainer />
    </NetworkProvider>
  );
}

export default App;
