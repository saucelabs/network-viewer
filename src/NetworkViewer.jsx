import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import NetworkProvider from './state/network/NetworkProvider';
import MainContainer from './Containers/MainContainer';
import Styles from './NetworkViewer.styles.scss';
import ThemeProvider from './state/theme/Context';

const contextClassNames = classNames.bind(Styles);

const NetworkViewer = ({
  file,
  data,
  fetchOptions,
  scrollTimeStamp,
  options,
  onPause,
  onRequestSelect,
  onResume,
  onReset,
  scrollRequestPosition,
  autoHighlightChange,
  onDataLoaded,
  onDataError,
  containerClassName,
}) => (
  <section className={contextClassNames('network-viewer', containerClassName)}>
    <ThemeProvider options={options}>
      <NetworkProvider
        autoHighlightChange={autoHighlightChange}
        data={data}
        fetchOptions={fetchOptions}
        file={file}
        onDataError={onDataError}
        onDataLoaded={onDataLoaded}
        onPause={onPause}
        onReset={onReset}
        onResume={onResume}
        scrollRequestPosition={scrollRequestPosition}
        scrollTimeStamp={scrollTimeStamp}
      >
        <MainContainer onRequestSelect={onRequestSelect} />
      </NetworkProvider>
    </ThemeProvider>
  </section>
);

NetworkViewer.propTypes = {
  autoHighlightChange: PropTypes.bool,
  containerClassName: PropTypes.string,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onRequestSelect: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  options: PropTypes.object,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkViewer.defaultProps = {
  autoHighlightChange: false,
  containerClassName: null,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onRequestSelect: () => {},
  onReset: null,
  onResume: null,
  options: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkViewer;
