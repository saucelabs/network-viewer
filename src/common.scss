@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
