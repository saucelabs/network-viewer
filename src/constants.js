export const VIEWER_FIELDS = [
  {
    key: 'status',
    name: 'Status',
  },
  {
    key: 'method',
    name: 'Method',
  },
  {
    key: 'domain',
    name: 'Domain',
  },
  {
    key: 'filename',
    name: 'File',
  },
  {
    key: 'type',
    name: 'Type',
  },
  {
    key: 'size',
    name: 'Size',
    unit: 'KB',
  },
];

export const DEFAULT_FILTER = {
  key: null,
  value: null,
};

export const FILTERS = [
  {
    name: 'All',
    filterBy: DEFAULT_FILTER,
  },
  {
    name: 'HTML',
    filterBy: {
      key: 'type',
      value: 'html',
    },
  },
  {
    name: 'CSS',
    filterBy: {
      key: 'type',
      value: 'css',
    },
  },
  {
    name: 'JS',
    filterBy: {
      key: 'type',
      value: 'javascript',
    },
  },
  {
    name: 'XHR',
    filterBy: {
      key: 'type',
      value: 'xhr',
    },
  },
  {
    name: 'Fonts',
    filterBy: {
      key: 'type',
      value: 'fonts',
    },
  },
  {
    name: 'Images',
    filterBy: {
      key: 'type',
      value: 'images',
    },
  },
];
