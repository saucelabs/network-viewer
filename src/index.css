html {
  height: 100%;
}

body {
  height: 100%;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

div:focus {
  outline: none;
}

#root {
  display: flex;
  flex-direction: column;
  height: 100%;
}
