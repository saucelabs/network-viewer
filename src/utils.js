export const getUrlInfo = (url) => {
  const urlInfo = new URL(url);
  const pathSplit = urlInfo.pathname.split('/');
  const fileName = pathSplit[pathSplit.length - 1].trim()
    ? pathSplit[pathSplit.length - 1] : pathSplit[pathSplit.length - 2];

  return {
    domain: urlInfo.host,
    filename: fileName + urlInfo.search,
    url: urlInfo.href,
  };
};

export const parseSize = ({ headers, content }) => {
  if (content && content.size) {
    return Number((content.size / 1024).toFixed(2));
  }

  const contentInfo = headers.find(({ name }) => ['content-length', 'Content-Length'].includes(name));
  if (!contentInfo) {
    return 0;
  }

  return Number((contentInfo.value / 1024).toFixed(2));
};

export const getContentType = (headers) => {
  const contentInfo = headers.find(({ name }) => ['content-type', 'Content-Type'].includes(name));
  if (!contentInfo) {
    return '';
  }
  const type = contentInfo.value.split(';')[0].split('/');
  return type.length > 1 ? type[1] : type[0];
};

export const getTimings = ({ startedDateTime, timings }, firstEntryTime) => ({
  ...timings,
  startTime: new Date(startedDateTime).getTime() - new Date(firstEntryTime).getTime(),
});

export const prepareViewerData = (entries) => {
  const firstEntryTime = entries[0].startedDateTime;
  const lastEntryTime = entries[entries.length - 1].startedDateTime;
  const data = entries
    .filter(entry => entry.response)
    .map((entry, index) => ({
      index,
      status: entry.response.status,
      method: entry.request.method,
      size: parseSize(entry.response),
      startedDateTime: entry.startedDateTime,
      type: getContentType(entry.response.headers),
      timings: getTimings(entry, firstEntryTime),
      ...getUrlInfo(entry.request.url),
    }));

  const totalNetworkTime = new Date(lastEntryTime).getTime()
    - new Date(firstEntryTime).getTime()
    + data[data.length - 1].timings.receive;
  return {
    totalNetworkTime,
    data,
  };
};

export const sortBy = (data, key, isAsc = true) => data.sort((prev, next) => {
  if (prev[key] < next[key]) {
    return isAsc ? -1 : 1;
  }
  if (prev[key] > next[key]) {
    return isAsc ? 1 : 1;
  }
  return 0;
});

export const filterData = (data, filter) => {
  const { search, key, value } = filter;
  const trimmedSearch = search.trim();

  return !trimmedSearch && !key
    ? data
    : data.filter((info) => {
      const isSearchMatched = trimmedSearch ? info.filename.includes(trimmedSearch) : true;
      const isFilterMatched = key ? info[key] === value : true;
      return isSearchMatched && isFilterMatched;
    });
};

export const actionsWrapper = (actions = {}) => (dispatch, state) => Object.keys(actions)
  .reduce((modifiedActions, type) => ({
    ...modifiedActions,
    [type]: actions[type](dispatch, state),
  }), {});
