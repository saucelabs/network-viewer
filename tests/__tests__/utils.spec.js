/* eslint no-useless-escape: 0 */
import { List } from 'immutable';

import * as utils from './../../src/utils';
import networkDataMock from './../__fixtures__/network.json';
import interceptErrorNetworkDataMock from './../__fixtures__/network_intercept_error.json';
import preparedMockData from './../__fixtures__/preparedData';

describe('utils', () => {
  it('getUrlInfo', () => {
    const url = 'https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2';
    expect(utils.getUrlInfo(url)).toMatchSnapshot();
  });

  it('getUrlInfo non standard url / resource identifier', () => {
    const resourceName = 'ResourceIdentifierString';
    const urlInfo = utils.getUrlInfo(resourceName);

    expect(urlInfo).toEqual({
      domain: 'N/A',
      filename: resourceName,
      url: resourceName,
    });
  });

  it('parseSize', () => {
    const { response } = networkDataMock.log.entries[0];
    expect(utils.parseSize(response)).toMatchSnapshot();
  });

  it('getContentType', () => {
    const entry = networkDataMock.log.entries[0];
    expect(utils.getContentType(entry)).toMatchSnapshot();

    const entryWithoutResourceType = {
      ...entry,
      _resourceType: null,
    };
    expect(utils.getContentType(entryWithoutResourceType)).toMatchSnapshot();
  });

  it('getTimings', () => {
    const firstEntryTime = networkDataMock.log.entries[0].startedDateTime;
    expect(utils.getTimings(networkDataMock.log.entries[3], firstEntryTime)).toMatchSnapshot();
  });

  it('prepareViewerData', () => {
    const { entries } = networkDataMock.log;
    expect(utils.prepareViewerData(entries)).toMatchSnapshot();
    expect(utils.prepareViewerData([])).toMatchSnapshot();
  });

  it('prepareViewerData when intercept error exists', () => {
    expect(utils.prepareViewerData(interceptErrorNetworkDataMock.log.entries)).toMatchSnapshot();
  });

  it('sortBy', () => {
    expect(utils.sortBy(preparedMockData, 'size').toJS()).toMatchSnapshot();
  });

  it('filterData', () => {
    expect(utils.filterData({
      data: preparedMockData,
      search: { name: 'url', value: 'e96c15f68c68' },
      filter: {},
      errorFilter: false,
    }).toJS()).toMatchSnapshot();
    expect(utils.filterData({
      data: preparedMockData,
      filter: { key: 'type', value: ['html'] },
      search: {},
      errorFilter: false,
    }).toJS()).toMatchSnapshot();
    expect(utils.filterData({
      data: preparedMockData,
      errorFilter: true,
      search: {},
      filter: {},
    }).toJS()).toMatchSnapshot();
  });

  it('getContent', () => {
    expect(utils.getContent({
      mimeType: 'application/json',
      text: '{\n  \"foo\": \"bar\"\n}',
    })).toMatchSnapshot();
    expect(utils.getContent({ content: 'cool' })).toMatchSnapshot();
  });

  it('prepareTooltipData', () => {
    const data = {
      blocked: 53.00000000000955,
      connect: 32,
      dns: 9,
      receive: 530.0000000000152,
      send: 5,
      ssl: 20,
      wait: 588.9999999999925,
      _blocked_queueing: 2.0000000000095497,
      startTime: 671,
    };
    expect(utils.prepareTooltipData(data)).toMatchSnapshot();
  });

  it('calcTotalTime', () => {
    const data = {
      blocked: 53.00000000000955,
      connect: 32,
      dns: 9,
      receive: 530.0000000000152,
      send: 5,
      ssl: 20,
      wait: 588.9999999999925,
      _blocked_queueing: 2.0000000000095497,
      startTime: 671,
    };
    expect(utils.calcTotalTime(data)).toMatchSnapshot();
  });

  it('calcChartAttributes', () => {
    const data = {
      blocked: 240.50000000000034,
      connect: 42,
      dns: 14,
      receive: 7045.500000000004,
      send: 0,
      ssl: 24,
      wait: 786.5000000000023,
      _blocked_queueing: 196.50000000000034,
      startTime: 672,
    };
    expect(utils.calcChartAttributes(data, '17256.999999999985')).toMatchSnapshot();
  });

  describe('formatSize', () => {
    it('converts Bytes', () => {
      expect(utils.formatSize(50)).toBe('50 B');
    });

    it('converts Kilobytes', () => {
      expect(utils.formatSize(90001)).toBe('87.9 KB');
    });

    it('converts Megabytes', () => {
      expect(utils.formatSize(4448576)).toBe('4.2 MB');
    });
  });

  describe('formatTime', () => {
    it('converts Millisecond', () => {
      expect(utils.formatTime(900)).toBe('900ms');
    });

    it('converts Seconds', () => {
      expect(utils.formatTime(9001)).toBe('9.01s');
    });

    it('converts Minutes', () => {
      expect(utils.formatTime(90001)).toBe('2m');
    });
  });

  it('getHeaders', () => {
    expect(utils.getHeaders({
      request: {
        headers: [{ name: 'b' }, { name: 'a' }],
      },
      response: {
        headers: [{ name: 'z' }, { name: 'y' }, { name: 'x' }],
      },
    })).toMatchSnapshot();
  });

  describe('roundOff', () => {
    it('should return correct value', () => {
      expect(utils.roundOff(50)).toBe(50);
      expect(utils.roundOff(50.1234)).toBe(50.1);
      expect(utils.roundOff(50.1634)).toBe(50.2);
      expect(utils.roundOff(50.1634, 2)).toBe(50.16);
    });
  });

  describe('getSummary', () => {
    it('should return correct value', () => {
      const data = new List([{
        transferredSize: 1343,
        uncompressedSize: 2400,
      }, {
        transferredSize: 2090,
        uncompressedSize: 3000,
      }, {
        transferredSize: 200,
        uncompressedSize: 650,
      }]);
      expect(utils.getSummary(data)).toMatchSnapshot();
    });
  });

  describe('getTotalTimeOfEntry', () => {
    it('should return totalTimeOfEntry', () => {
      expect(utils.getTotalTimeOfEntry(networkDataMock.log.entries[0])).toMatchSnapshot();
    });

    it('should work without _blocked_queueing in timings', () => {
      // eslint-disable-next-line camelcase
      const { _blocked_queueing, ...timings } = networkDataMock.log.entries[0].timings;
      const entry = { ...networkDataMock.log.entries[0], timings };
      expect(utils.getTotalTimeOfEntry(entry)).toMatchSnapshot();
    });
  });

  it('findIndexNearTimestamp', () => {
    expect(utils.findIndexNearTimestamp(preparedMockData, 1571042841141)).toMatchSnapshot();
  });

  it('findIndexBeforeTimestamp', () => {
    expect(utils.findIndexBeforeTimestamp(preparedMockData, 1571042835643)).toMatchSnapshot();
  });

  it('findIndexAfterTimestamp', () => {
    expect(utils.findIndexAfterTimestamp(preparedMockData, 1571042835643)).toMatchSnapshot();
  });

  it('getStatusClass', () => {
    expect(utils.getStatusClass({ status: 200 })).toBe('info');
    expect(utils.getStatusClass({ status: 503 })).toBe('error');
    expect(utils.getStatusClass({ status: 0, error: 'ERR' })).toBe('error');
    expect(utils.getStatusClass({ status: 0 })).toBe('pending');
  });

  it('formatValue', () => {
    expect(utils.formatValue('status', 200)).toBe(200);
    expect(utils.formatValue('status', 0, '', { error: 'ERR' })).toBe('(failed)');
    expect(utils.formatValue('status', 0)).toBe('Pending');
    expect(utils.formatValue('status', 0)).toBe('Pending');
  });

  it('getInterceptError', () => {
    expect(utils.getInterceptError({ response: { _error: 'ERR_TIMED_OUT' } })).toBe('ERR_TIMED_OUT');
    expect(utils.getInterceptError({ response: { } })).toBe(null);
  });

  it('parseRequestPayload', () => {
    expect(utils.parseRequestPayload("{'name':'foo'}")).toMatchSnapshot();
    expect(utils.parseRequestPayload('foo')).toBe('foo');
  });
});
