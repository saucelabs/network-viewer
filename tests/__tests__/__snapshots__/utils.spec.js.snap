// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils calcChartAttributes 1`] = `
Array [
  Object {
    "fill": "#ccc",
    "key": "queueing",
    "width": "1.1386683664599901%",
    "x": "3.8940719707944633%",
  },
  Object {
    "fill": "#A1000C",
    "key": "blocked",
    "width": "1.393637364547723%",
    "x": "5.032740337254453%",
  },
  Object {
    "fill": "#DCC9E5",
    "key": "dns",
    "width": "0.08112649939155132%",
    "x": "6.426377701802176%",
  },
  Object {
    "fill": "#E78057",
    "key": "ssl",
    "width": "0.13907399895694514%",
    "x": "6.507504201193727%",
  },
  Object {
    "fill": "#DB8553",
    "key": "connect",
    "width": "0.24337949817465396%",
    "x": "6.646578200150673%",
  },
  Object {
    "fill": "#7CA0BF",
    "key": "wait",
    "width": "4.557570840818236%",
    "x": "6.889957698325326%",
  },
  Object {
    "fill": "#65B955",
    "key": "receive",
    "width": "40.826910818798225%",
    "x": "11.447528539143562%",
  },
]
`;

exports[`utils calcTotalTime 1`] = `1238.0000000000173`;

exports[`utils filterData 1`] = `
Array [
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": 1571042831641,
    "status": 200,
    "time": 53.14699999871664,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
]
`;

exports[`utils filterData 2`] = `
Array [
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": 1571042831592,
    "status": 200,
    "time": 35.07000001263805,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": 1571042831641,
    "status": 200,
    "time": 40.115000010700896,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": 1571042831641,
    "status": 200,
    "time": 53.14699999871664,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": 1571042831643,
    "status": 200,
    "time": 51.217000000178814,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": 1571042831644,
    "status": 200,
    "time": 53.195000014966354,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;

exports[`utils getContent 1`] = `"{\\"foo\\":\\"bar\\"}"`;

exports[`utils getContent 2`] = `undefined`;

exports[`utils getContentType 1`] = `"html"`;

exports[`utils getTimings 1`] = `
Object {
  "_blocked_queueing": 11.349999986123294,
  "blocked": 13.026999986123293,
  "connect": -1,
  "dns": -1,
  "receive": 1.951000012923032,
  "send": 0.21299999999999986,
  "ssl": -1,
  "startTime": 51,
  "wait": 36.02600000113249,
}
`;

exports[`utils getUrlInfo 1`] = `
Object {
  "domain": "developer.mozilla.org",
  "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
}
`;

exports[`utils parseSize 1`] = `135.04`;

exports[`utils prepareTooltipData 1`] = `
Object {
  "_blocked_queueing": "2.00 ms",
  "blocked": "53.00 ms",
  "connect": "32.00 ms",
  "dns": "9.00 ms",
  "queuedAt": "671.00 ms",
  "receive": "530.00 ms",
  "send": "5.00 ms",
  "ssl": "20.00 ms",
  "startTime": "671.00 ms",
  "startedAt": "673.00 ms",
  "totalTime": "1.24 s",
  "wait": "589.00 ms",
}
`;

exports[`utils prepareViewerData 1`] = `
Object {
  "data": Array [
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "slice",
      "index": 0,
      "method": "GET",
      "size": 135.04,
      "startedDateTime": 1571042831592,
      "status": 200,
      "time": 35.07000001263805,
      "timings": Object {
        "_blocked_queueing": 4.382999992230907,
        "blocked": 5.496999992230907,
        "connect": -1,
        "dns": -1,
        "receive": 3.252000024076551,
        "send": 0.2819999999999998,
        "ssl": -1,
        "startTime": 0,
        "wait": 26.03899999633059,
      },
      "type": "document",
      "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
    },
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
      "index": 1,
      "method": "GET",
      "size": 33.02,
      "startedDateTime": 1571042831641,
      "status": 200,
      "time": 40.115000010700896,
      "timings": Object {
        "_blocked_queueing": 12.528999999631196,
        "blocked": 13.952999999631196,
        "connect": -1,
        "dns": -1,
        "receive": 7.124000025214627,
        "send": 0.2370000000000001,
        "ssl": -1,
        "startTime": 49,
        "wait": 18.80099998585507,
      },
      "type": "font",
      "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    },
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
      "index": 2,
      "method": "GET",
      "size": 33.27,
      "startedDateTime": 1571042831641,
      "status": 200,
      "time": 53.14699999871664,
      "timings": Object {
        "_blocked_queueing": 13.034000003244728,
        "blocked": 14.968000003244727,
        "connect": -1,
        "dns": -1,
        "receive": 1.9629999878816307,
        "send": 0.40400000000000014,
        "ssl": -1,
        "startTime": 49,
        "wait": 35.81200000759028,
      },
      "type": "font",
      "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    },
    Object {
      "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
      "domain": "developer.mozilla.org",
      "filename": "react-header.2a659f6ad280.css",
      "index": 3,
      "method": "GET",
      "size": 2.76,
      "startedDateTime": 1571042831643,
      "status": 200,
      "time": 51.217000000178814,
      "timings": Object {
        "_blocked_queueing": 11.349999986123294,
        "blocked": 13.026999986123293,
        "connect": -1,
        "dns": -1,
        "receive": 1.951000012923032,
        "send": 0.21299999999999986,
        "ssl": -1,
        "startTime": 51,
        "wait": 36.02600000113249,
      },
      "type": "stylesheet",
      "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
    },
    Object {
      "body": "",
      "domain": "developer.mozilla.org",
      "filename": "react-mdn.c66d8d89d98e.css",
      "index": 4,
      "method": "GET",
      "size": 142.88,
      "startedDateTime": 1571042831644,
      "status": 200,
      "time": 53.195000014966354,
      "timings": Object {
        "_blocked_queueing": 11.250000010477379,
        "blocked": 13.423000010477379,
        "connect": -1,
        "dns": -1,
        "receive": 1.2719999940600246,
        "send": 0.5390000000000001,
        "ssl": -1,
        "startTime": 52,
        "wait": 37.96100001042895,
      },
      "type": "stylesheet",
      "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
    },
  ],
  "totalNetworkTime": 53.271999994060025,
}
`;

exports[`utils sortBy 1`] = `
Array [
  Object {
    "body": ".logo{display:block;grid-area:I;width:200px;height:44px;margin-right:24px}",
    "domain": "developer.mozilla.org",
    "filename": "react-header.2a659f6ad280.css",
    "index": 3,
    "method": "GET",
    "size": 2.76,
    "startedDateTime": 1571042831643,
    "status": 200,
    "time": 51.217000000178814,
    "timings": Object {
      "_blocked_queueing": 11.349999986123294,
      "blocked": 13.026999986123293,
      "connect": -1,
      "dns": -1,
      "receive": 1.951000012923032,
      "send": 0.21299999999999986,
      "ssl": -1,
      "startTime": 51,
      "wait": 36.02600000113249,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-header.2a659f6ad280.css",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
    "index": 1,
    "method": "GET",
    "size": 33.02,
    "startedDateTime": 1571042831641,
    "status": 200,
    "time": 40.115000010700896,
    "timings": Object {
      "_blocked_queueing": 12.528999999631196,
      "blocked": 13.952999999631196,
      "connect": -1,
      "dns": -1,
      "receive": 7.124000025214627,
      "send": 0.2370000000000001,
      "ssl": -1,
      "startTime": 49,
      "wait": 18.80099998585507,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Regular.subset.bbc33fb47cf6.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "ZillaSlab-Bold.subset.e96c15f68c68.woff2",
    "index": 2,
    "method": "GET",
    "size": 33.27,
    "startedDateTime": 1571042831641,
    "status": 200,
    "time": 53.14699999871664,
    "timings": Object {
      "_blocked_queueing": 13.034000003244728,
      "blocked": 14.968000003244727,
      "connect": -1,
      "dns": -1,
      "receive": 1.9629999878816307,
      "send": 0.40400000000000014,
      "ssl": -1,
      "startTime": 49,
      "wait": 35.81200000759028,
    },
    "type": "font",
    "url": "https://developer.mozilla.org/static/fonts/locales/ZillaSlab-Bold.subset.e96c15f68c68.woff2",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "slice",
    "index": 0,
    "method": "GET",
    "size": 135.04,
    "startedDateTime": 1571042831592,
    "status": 200,
    "time": 35.07000001263805,
    "timings": Object {
      "_blocked_queueing": 4.382999992230907,
      "blocked": 5.496999992230907,
      "connect": -1,
      "dns": -1,
      "receive": 3.252000024076551,
      "send": 0.2819999999999998,
      "ssl": -1,
      "startTime": 0,
      "wait": 26.03899999633059,
    },
    "type": "document",
    "url": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice",
  },
  Object {
    "body": "",
    "domain": "developer.mozilla.org",
    "filename": "react-mdn.c66d8d89d98e.css",
    "index": 4,
    "method": "GET",
    "size": 142.88,
    "startedDateTime": 1571042831644,
    "status": 200,
    "time": 53.195000014966354,
    "timings": Object {
      "_blocked_queueing": 11.250000010477379,
      "blocked": 13.423000010477379,
      "connect": -1,
      "dns": -1,
      "receive": 1.2719999940600246,
      "send": 0.5390000000000001,
      "ssl": -1,
      "startTime": 52,
      "wait": 37.96100001042895,
    },
    "type": "stylesheet",
    "url": "https://developer.mozilla.org/static/build/styles/react-mdn.c66d8d89d98e.css",
  },
]
`;
